temporal/error.go

```diff
@@ -166,6 +168,80 @@ var (
 	ErrSkipScheduleUpdate = internal.ErrSkipScheduleUpdate
 )
 
+// ApplicationErrorAttributes should be used to set all the desired attributes of a new ApplicationError
+// To get a new instance use ErrorAttributes function
+type ApplicationErrorAttributes struct {
+	message     string
+	errType     string
+	isRetryable bool
+	cause       error
+	details     []interface{}
+}
+
+type errorAttributes = ApplicationErrorAttributes
+
+// Produce a new instance compliant with ApplicationErrorAttributes
+func ErrorAttributes() ApplicationErrorAttributes {
+	attrs := ApplicationErrorAttributes{isRetryable: true}
+	return attrs
+}
+
+func (b ApplicationErrorAttributes) WithMessage(message string) ApplicationErrorAttributes {
+	b.message = message
+	return b
+}
+
+func (b ApplicationErrorAttributes) WithErrorType(errType string) ApplicationErrorAttributes {
+	b.errType = errType
+	return b
+}
+
+func (b ApplicationErrorAttributes) WithRetryable(isRetryable bool) ApplicationErrorAttributes {
+	b.isRetryable = isRetryable
+	return b
+}
+
+func (b ApplicationErrorAttributes) WithDetails(details ...interface{}) ApplicationErrorAttributes {
+	b.details = details
+	return b
+}
+
+func (b ApplicationErrorAttributes) WithCause(cause error) ApplicationErrorAttributes {
+	b.cause = cause
+	return b
+}
+
+func (b ApplicationErrorAttributes) asErrorAttributes() internal.ErrorAttributes {
+	return errorAttributes(b)
+}
+
+func (b errorAttributes) Message() string {
+	return b.message
+}
+
+func (b errorAttributes) ErrType() string {
+	return b.errType
+}
+
+func (b errorAttributes) IsRetryable() bool {
+	return b.isRetryable
+}
+
+func (b errorAttributes) Cause() error {
+	return b.cause
+}
+
+func (b errorAttributes) Details() []interface{} {
+	return b.details[:]
+}
+
+// NewApplicationErrorWithExtraInformation creates new instance of *ApplicationError type, all the attributes of the
+// newly created error could be controlled through instance of ApplicationErrorAttributes see: ErrorAttributes
+// This function also receives some extra information. See activity.ActivityExtraInformation for details.
+func NewApplicationErrorWithExtraInformation(attributes ApplicationErrorAttributes, extras ExtraInformation) error {
```

Comment:
I think this should look like https://github.com/temporalio/sdk-go/pull/1319#discussion_r1425330300:

```
type ApplicationErrorOptions struct {
    Details []any
    Cause error
    NonRetryable bool
    NextAttemptRetryDelay time.Duration
}

func NewApplicationErrorWithOptions(message, errType string, options ApplicationErrorOptions) error
```

You don't need all the extra indirection.

Reply:
Yes please. This is way more inline with the other SDK APIs.